### ISSUE TITLE
[Tempo] - Overrides section changes

### ISSUE THREAD

**Issue by bmgante** [grafana/helm-charts#2802]
Hi,
Wondering to upgrade to tempo 2.3.0 and noticed the following section has changed:
![image](https://github.com/grafana/helm-charts/assets/27358817/7c98bdec-e340-4107-8adf-5223a586f854)

The helm chart which is updated seems not to reflect this change, or am i wrong?
```
  overrides:
    {{- toYaml .Values.global_overrides | nindent 2 }}
    {{- if .Values.metricsGenerator.enabled }}
    metrics_generator_processors:
    {{- range .Values.global_overrides.metrics_generator_processors }}
    - {{ . }}
    {{- end }}
    {{- end }}
```

My values.yaml needs to be changed? It is currently as below
```
# Global overrides
global_overrides:
  per_tenant_override_config: /runtime-config/overrides.yaml
  metrics_generator_processors:
    - service-graphs
    - span-metrics
```
Thanks

**Comment by edgarkz** [comment]
+1. At this moment the new style is not working using helm value. 
As a temporary solution - add manual section:
`        overrides: |   
            overrides:
              "*":
                ingestion_burst_size_bytes: 50000000
                ingestion_rate_limit_bytes: 35000000
`

**Comment by bmgante** [comment]
What if changing the values yaml this way, would it work?

From:
```
  overrides:
    {{- toYaml .Values.global_overrides | nindent 2 }}
    {{- if .Values.metricsGenerator.enabled }}
    metrics_generator_processors:
    {{- range .Values.global_overrides.metrics_generator_processors }}
    - {{ . }}
    {{- end }}
    {{- end }}

...
global_overrides:
  per_tenant_override_config: /runtime-config/overrides.yaml
  #metrics_generator_processors: []
  metrics_generator_processors:
    - service-graphs
    - span-metrics
```

To:
```
  overrides:
  defaults:
    {{- toYaml .Values.global_overrides | nindent 2 }}
    {{- if .Values.metricsGenerator.enabled }}
    metrics_generator:
    {{- range .Values.global_overrides.metrics_generator.processors }}
    - {{ . }}
    {{- end }}
    {{- end }}

...
global_overrides:
  per_tenant_override_config: /runtime-config/overrides.yaml
  #metrics_generator_processors: []
  metrics_generator:
    processors:
      - service-graphs
      - span-metrics
      - local-blocks
```

**Comment by bmgante** [comment]
@edgarkz could you please elaborate a bit more on how to apply your manual workaround to have the helm chart working fine with 2.3.x tempo version and the new override syntax?

**Comment by AlexDCraig** [comment]
Yeah we definitely need this, looks like https://github.com/grafana/helm-charts/pull/2825 is necessary. If you're like me and have some global overrides it may make sense to wait until this patch is released

**Comment by edgarkz** [comment]
> @edgarkz could you please elaborate a bit more on how to apply your manual workaround to have the helm chart working fine with 2.3.x tempo version and the new override syntax?

It doesn't work with new syntax.. I have added old syntax overrides to make it work in tempo 2.3
      values: |-
        overrides: |   
          overrides:
            "*":

**Comment by pantuza** [comment]
## TL;DR
On the Helm Chart `tempo-distributed` instead of adding a parameter `overrides`, use `global_overrides` as a workaround:
```yaml
global_overrides:
  defaults:
    ingestion:
      rate_limit_bytes: 32000000 # 32MB
      burst_size_bytes: 48000000 # 48MB
      max_traces_per_user: 50000
```
> This is a workaround. The global_overrides isn't the right place but since the template outputs its content to the right final overrides block, you can use it by now.
 
## Why the issue is happening?
The Grafana Tempo [overrides documentation](https://grafana.com/docs/tempo/latest/configuration/#overrides) is correct! If  your final yaml file ends up with a `overrides` block, it will work. 
Although, `distributed-tempo` Helm Chart is where the issue happens.  Since version [1.8.0](https://github.com/grafana/helm-charts/tree/main/charts/tempo-distributed#from-chart-versions--180) of this Helm Chart it is using a new block called `overrides` that accepts a string. Even if you provide the proper string content it will fail simply because the Helm template isn't adding this block on the final yaml file. 

Basically, the helm read the content of that block and add to a variable `tempo.OverridesConfig` ([reference](https://github.com/grafana/helm-charts/blob/080650738b427aaee657f3f42af96a9fc45b8d42/charts/tempo-distributed/templates/_helpers.tpl#L190)).
Then it creates a ConfigMap that uses the content of this variable as a new file `overrides.yaml` ([reference](https://github.com/grafana/helm-charts/blob/080650738b427aaee657f3f42af96a9fc45b8d42/charts/tempo-distributed/templates/configmap-runtime.yaml#L10)).
And here comes the issue: when it generates the final yaml (`tempo.yaml`) it does not add the overrides block ([reference](https://github.com/grafana/helm-charts/blob/080650738b427aaee657f3f42af96a9fc45b8d42/charts/tempo-distributed/templates/configmap-tempo.yaml#L21)). 
It lets its content on that external file called `overrides.yaml`. 

## Possible solutions
I think in [this block](https://github.com/grafana/helm-charts/blob/080650738b427aaee657f3f42af96a9fc45b8d42/charts/tempo-distributed/values.yaml#L1252) the Helm Chart should not only add `global_overrides` but also `overrides` block as its content. 
This would output the overrides content into the final yaml file. Thus, solving the issue because Tempo binary will read the expected `overrides` block properly.

**Comment by batazor** [comment]
At the moment I am unable to enable **service graph**, any tips on how we can do this?

How can I activate `enable_virtual_node_label`?

<img width="1790" alt="Screenshot 2024-07-21 at 11 09 30" src="https://github.com/user-attachments/assets/45ea5e81-17c7-4ce7-98e7-1341b20728fb">

**Comment by ThomasVitale** [comment]
The `global_overrides `doesn't seem to work anymore (chart version 1.15.2). I succeeded in correctly passing `overrides `for the service graph as follows.

```yaml
tempo:
  structuredConfig:
    overrides:
      defaults:
        metrics_generator:
          processors: [service-graphs, span-metrics, local-blocks]
          generate_native_histograms: both
```

**Comment by chenlujjj** [comment]
Below is part of my value file which makes the metrics generator work
```
  # Global overrides
  global_overrides:
    per_tenant_override_config: /runtime-config/overrides.yaml
    defaults:
      metrics_generator:
        processors: [service-graphs, span-metrics]


  # Per tenants overrides
  overrides: {}
```
chart version is 1.16.2

@ThomasVitale @batazor

**Comment by shinebayar-g** [comment]
Another confused user here. After spending almost an hour reading everywhere, I think the intended usage was:

```yaml
overrides:
    '*':
        metrics_generator:
            processors: ['service-graphs', 'span-metrics']
```

Helm chart version: `grafana/tempo-distributed: 1.21.0`

But as soon as config is applied, I'm getting rate limited errors on the Alloy.

```log
2024-11-07T03:09:59.590772917Z stderr F ts=2024-11-07T03:09:59.590614967Z level=error msg="Exporting failed. Dropping data." component_path=/ component_id=otelcol.exporter.otlp.tempo error="not retryable error: Permanent error: rpc error: code = ResourceExhausted desc = RATE_LIMITED: ingestion rate limit (local: 0 bytes, global: 0 bytes) exceeded while adding 13541 bytes for user single-tenant" dropped_items=10
```

It seems that override config doesn't merge with the global config, so you have to reconfigure everything again which is crazy and error prone. I don't want to go down this rabbit hole and figure out which configs should be restored in the `overrides:`.

If my analysis are correct, you **_should not_** use `overrides:` config in the helm chart unless you know what you're doing.
It's not doing overrides as it suggests, it's doing replacements. Opposite of how `values.yaml` works in helm chart.

Fortunately this config worked for me. Service Graph is populated, Alloy no longer getting rate limited errors.

```yaml
global_overrides:
    defaults:
        metrics_generator:
            processors: ['service-graphs', 'span-metrics']
```

@grafana can we improve this situation?

Related issues:
- https://github.com/grafana/tempo/issues/3855
- https://github.com/grafana/tempo/issues/3795

**Comment by knylander-grafana** [comment]
Hi there! I noticed that we had a bunch of issues open for the overrides settings for the tempo-distributed Helm chart. We have two updates that add more doc for overrides: 
* https://github.com/grafana/helm-charts/pull/3468
* https://github.com/grafana/tempo/pull/4415

Please let me know if this addresses the issue.

**Comment by lkiii** [comment]
I am struggling to understand how to use the `metricsGenerator` with `tempo` (not distributed) helm chart. From the docs and all the explanations it looks like I would need to have values like:

```yaml
tempo:
  metricsGenerator:
    enabled: true
  global_overrides:
    metrics_generator_processors:
      - local-blocks
```
The issue is with https://github.com/grafana/helm-charts/blob/dfeecb93fff3057d8690c6c13cd2cbe62d08c55d/charts/tempo/values.yaml#L171-L174

I do not see any way extending `metrics_generator_processors` as it is being created on condition `{{- if .Values.tempo.metricsGenerator.enabled }}`


and the result of the config map has:

```yaml
    overrides:
          metrics_generator_processors:
          - local-blocks
          per_tenant_override_config: /conf/overrides.yaml
          metrics_generator_processors:
          - 'service-graphs'
          - 'span-metrics'
```

Which breaks tempo.

**Comment by skorzhevsky** [comment]
Same question as above, but for tempo-distributed. How is it possible to enable "local-blocks" and set its parameter (e.g. filter_server_spans: false)? Also, disable other processors. Thank you.

**Comment by rlex** [comment]
for tempo-distributed i use following:
```yaml
global_overrides:
   metrics_generator_processors:
       - service-graphs
       - local-blocks
```
which gets expanded to following in configmap:
```yaml
overrides:
  metrics_generator_processors:
  - service-graphs
  - local-blocks
  per_tenant_override_config: /runtime-config/overrides.yaml
```
this works for me for servicemaps, but i still get deprecation warning

**Comment by bobafettfrom** [comment]
> At the moment I am unable to enable **service graph**, any tips on how we can do this?
> 
> How can I activate `enable_virtual_node_label`?
> 
> <img alt="Screenshot 2024-07-21 at 11 09 30" width="1790" src="https://private-user-images.githubusercontent.com/1643797/350749927-45ea5e81-17c7-4ce7-98e7-1341b20728fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg3NDA2NzYsIm5iZiI6MTczODc0MDM3NiwicGF0aCI6Ii8xNjQzNzk3LzM1MDc0OTkyNy00NWVhNWU4MS0xN2M3LTRjZTctOThlNy0xMzQxYjIwNzI4ZmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDVUMDcyNjE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODI1Yjg2NmQzZDU3ZjJmYzViYmRjMzc5ODAyZDc2NWFlYmUwMmMzOTA0YTdkNmQzYTFhYTRjY2QyNjhiYWFiNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GV5WEfgHq8Hz0mVZugQYq48zkpDlGnIAg0Ge3AqIVpk">

Can some help?

**Comment by jlcrow** [comment]
I gave up on upgrading Tempo, last attempt to upgrade left all pods in a state where they never start, never got beyond the config changes, had to restore the entire namespace from backup.
